### PR TITLE
Refactoring tests in the 'ApplicationFacadeTest' class

### PR DESCRIPTION
**Pull request by wassoares**
- Removed static imports;
- Addition of comment of the test cases and the related standardization in the body of the tests;
- Performing refactoring of some tests;

### CODE REVIEW

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 1]
```diff
@@ -12,1701 +12,2209 @@
...
+                .thenReturn(new SynchronizedDoublyLinkedList());
+
+        Mockito.doNothing().when(databaseManager).add(Matchers.any(Order.class));
+        Mockito.doNothing().when(databaseManager).update(Matchers.any(Order.class));
+
+        PowerMockito.mockStatic(DatabaseManager.class);
+        BDDMockito.given(DatabaseManager.getInstance()).willReturn(databaseManager);
+
+        this.orderController = Mockito.spy(new OrderController());
+        this.application = Mockito.spy(ApplicationFacade.getInstance());
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Please, check if it we really need to use a spies. If not, avoid using them.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 2]
```diff
@@ -12,1701 +12,2209 @@
...
-        expectedList.add("fake-order-id");
+        expectedList.add(FAKE_ORDER_ID);
 
         ComputeOrder computeOrder = Mockito.mock(ComputeOrder.class);
         Mockito.doReturn(expectedList).when(computeOrder).getNetworksId();
 
+        // exercise
         this.application.changeNetworkOrderIdsToNetworInstanceIds(computeOrder);
 
+        // verify
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I think we cannot mock the return of the computerOrder.getNetworkId. This is the behaviour we want to check. One option is to not use a mock for it or to use and to check we called the set method

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 3]
```diff
@@ -12,1701 +12,2209 @@
...
-		Order order = createComputeOrder();
-		OrderStateTransitioner.activateOrder(order);
+        // verify
+        Assert.assertEquals(OrderState.CLOSED, order.getOrderState());
+    }
+
+    // test case: When try calling the deleteCompute() method without user authentication, it must
+    // throw UnauthenticatedUserException and the Order remains in the same state.
+    @Test
+    public void testDeleteComputeOrderUnathenticated() throws Exception {
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I think we can improve the name of this test. It seems we are testing the forward of exception

Comment (thiagomanel) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: ... to me, it was hard to understand because the name has "Unathenticated". We are doing authorization, in fact.

Also, there is a typo in the name.

Comment (thiagomanel) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: ... in addition to that, the method do no call authenticate. Which indicates we are missing verifies in the code.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 4]
```diff
@@ -12,1701 +12,2209 @@
...
+        Mockito.doReturn(order).when(this.orderController).getOrder(Mockito.anyString(),
+                Mockito.any(FederationUser.class), Mockito.any(InstanceType.class));
 
-		Mockito.doNothing().when(this.aaaController).authorize(Mockito.any(FederationUser.class),
-				Mockito.any(Operation.class), Mockito.any(Order.class));
+        Mockito.doNothing().when(this.aaaController).authorize(Mockito.any(FederationUser.class),
+                Mockito.any(Operation.class), Mockito.any(Order.class));
+
+        Mockito.doThrow(new UnauthenticatedUserException()).when(this.aaaController)
+                .authenticate(Mockito.anyString());
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: why mocking this, if we do not call authenticate in the method under test?

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 5]
```diff
@@ -12,1701 +12,2209 @@
...
 
-		try {
-			String federationTokenValue = "";
-			this.application.deleteCompute(order.getId(), federationTokenValue);
-			Assert.fail();
-		} catch (UnauthenticatedUserException e) {
-			Assert.assertEquals(OrderState.OPEN, order.getOrderState());
-		}
-	}
+    // test case: When try calling the deleteCompute() method without a token authentication, it
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I think we should improve the description and the code. It seems to me we are testing the facade behaviour when the controller.getFederationUse throws an exception. This is unrelated to "without a token authentication"

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 6]
```diff
@@ -12,1701 +12,2209 @@
...
+            // verify
+            Assert.assertEquals(OrderState.OPEN, order.getOrderState());
+        }
+    }
 
-		Order order = createComputeOrder();
-		OrderStateTransitioner.activateOrder(order);
+    // test case: When try calling the deleteCompute() method performing an operation without
+    // authorization, it must throw UnauthorizedRequestException and the Order remains in the same
+    // state.
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: improve the description of the test case to reflect what the test do (or change the test to address the description). see my previous comments

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 7]
```diff
@@ -12,1701 +12,2209 @@
...
+            this.application.deleteCompute(order.getId(), FEDERATION_TOKEN_VALUE);
+            Assert.fail();
+        } catch (UnauthorizedRequestException e) {
+            // verify
+            Assert.assertEquals(OrderState.OPEN, order.getOrderState());
+        }
+    }
 
-		Mockito.doReturn(order.getFederationUser()).when(this.aaaController).getFederationUser(Mockito.anyString());
+    // test case: When calling the createCompute() method the Order passed by parameter must return
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I did not understand the case. Are we trying to check the order state was correctly assigned after the exercise? If so, what was the state before the exercise?

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 8]
```diff
@@ -12,1701 +12,2209 @@
...
-		}
-	}
+        Mockito.doThrow(new UnauthenticatedUserException()).when(this.aaaController)
+                .authenticate(Mockito.anyString());
 
-	@Test
-	public void testCreateComputeOrderTokenUnauthenticated() throws Exception {
-		ComputeOrder order = createComputeOrder();
+        Mockito.doReturn(order.getFederationUser()).when(this.aaaController)
+                .getFederationUser(Mockito.anyString());
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I think there is something wrong in the mock setup. The point is: how can this order.getFederationUser be called if we are going to throw an exception earlier? This is a symptom we are missing verifies. Also the description of the test case is not code, see the comment I made before.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 9]
```diff
@@ -12,1701 +12,2209 @@
...
 
-		try {
-			String federationTokenValue = "";
-			this.application.createCompute(order, federationTokenValue);
-			Assert.fail();
-		} catch (UnauthenticatedUserException e) {
-			Assert.assertNull(order.getOrderState());
-		}
-	}
+        Mockito.doThrow(new UnauthenticatedUserException()).when(this.aaaController)
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: same problems as before

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 10]
```diff
@@ -12,1701 +12,2209 @@
...
+            Assert.fail();
+        } catch (UnauthenticatedUserException e) {
+            // verify
+            Assert.assertNull(order.getOrderState());
+        }
+    }
 
-		Mockito.doNothing().when(this.aaaController).authorize(Mockito.any(FederationUser.class), Mockito.any(Operation.class),
-				Mockito.any(Order.class));
+    // test case: When try calling the createCompute() method with a Federation User without
```
Comment (thiagomanel) [inline src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: same problems as before

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 1 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: The spy on line 76 can be removed without affecting the code.
however, at line 75 its removal would cause failure in 12 of the 81 tests, and then I'll keep it.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 2 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Will be fixed at the next pull request...

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java [hunk 11]
```diff
@@ -12,1701 +12,2209 @@
...
 
-	@Test
-	public void testDeleteComputeOrderUnathenticated() throws Exception {
-		Mockito.doThrow(new UnauthenticatedUserException()).when(this.aaaController).authenticate(Mockito.anyString());
+        // exercise
+        this.application.deleteCompute(order.getId(), FEDERATION_TOKEN_VALUE);
 
-		Order order = createComputeOrder();
-		OrderStateTransitioner.activateOrder(order);
+        // verify
```
[the remark this answers is not in the document]
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Will be fixed at the next pull request...

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 3 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Modified the test method name for testDeleteComputeOrderWithoutAuthentication().

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 4 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: The 'deleteCompute()' method calls 'deleteOrder()' from the 'ApplicationFacade' class. This method 'deleteOrder()' before delegating its effective action to remove the order, requires the AaController class to authenticate the 'federationTokenValue'. And apparently 'Mock' is used to catch the exception thrown in the 'authenticate()' method of this class... 
But for me, this should only be tested in the 'AaControllerTest' class.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 6 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Fixed the name and description of the test case.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 7 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: Yes, the previous 'OrderState' was 'null'. Anyway, I fixed the description of the test case.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 8 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I will try to modify.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 9 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I will try to modify.

File: src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java (hunk 10 above)
Comment (wassoares) [reply src/test/java/org/fogbowcloud/manager/core/ApplicationFacadeTest.java]: I will try to modify.
